airflow/cli/commands/provider_command.py

```diff
@@ -17,7 +17,7 @@
 """Providers sub-commands."""
 from __future__ import annotations
 
-import re
+import re2 as re
```

Comment:
if there is a functional difference between `re` and `re2` we are bound to end up in issues.

Reply:
re is vulnerable attacks. re2 does not allow exponential backtracking.

You can find more info here:
https://lists.apache.org/thread/lytmbn1xf9vwgwfwgp4vrm3vshn8p1tm

Reply:
I think the solution is vague then in this case and a (innocuous looking) functional commit message to show that it addresses the core of the problem is required. What does re2 solve that re doesn't and we cannot do otherwise and why?

Reply:
Note: I think the actual issue is that we are trusting user input here and re2 seems on the surface right now to just to be band-aid and not addressing the core of the problem. But maybe you can elaborate?

Reply:
Fallback is explicit in `pyre2` (we are not using it), but I didn't find any mention of it in python `google-re2` bindings. I tried a few unsupported regexp and it threw errors straight away. I believe there is no fallback, this was discussed on the issue. re2 solves the ReDos problem with linear time regexp engine.

I don't have newer arguments than what has already been discussed in the github issue. (backtracking problem, pyre2 vs google-re2, fallback etc.).

As it is not yet released with a patch for this I purposely didn't give too many details here.

If needed can we follow up on the `airflow-s` issue, or security mailing list ?

Reply:
> Note: I think the actual issue is that we are trusting user input here and re2 seems on the surface right now to just to be band-aid and not addressing the core of the problem 

Not really. I think it's not a band-aid, using rgular expression is part of our API specification, so we cannot really remove it unless we have a very good reason (and it's actually useful). So solving a potential way how you could (mostly accidentally) trigger the situation where it it will take a lot of time is the right approach - we do not want to remove the functionality there. Moreover - since we will already have the google-re2 dependency (which BTW is proven and battle tested because it is used internally in `go` language), we can use the opportunity to use it elsewhere whre we use regular expressions and protect other pleaces. 

> Also re2 fallsback to re if it doesn't know how to handle the regex. But maybe you can elaborate?

The fallback is a mechanism for another library, It was a mistake to mention it.

Reply:
Yep I stand corrected on mentioning the fallback mechanism. I had to find out myself as there was so little detail in the commit message.

The commit message could read:

"Use linear time regular expressions

The standard regexp library can consume > O(n) in certain circumstances. The re2 library does not have this issue.
"

Which clarifies, but doesnt give away the issue. 

I do not fully agree with your assessment @potiuk that we want to keep that. We are trusting user input here and regexp engines are notorious to have issues. Imho the root cause is trusting user input and that is what probably should be probably. This is a workaround still.